crates/datafusion_ext/src/runtime/group_pull_up.rs

```diff
@@ -23,6 +24,54 @@ impl RuntimeGroupPullUp {
     pub fn new() -> Self {
         RuntimeGroupPullUp {}
     }
+
+    /// if a union consists of multiple runtimes, we can rewrite it as 2 groupings
+    /// union[remote, local, remote, remote] -> [union[union[remote, remote, remote], local]]
+    fn rewrite_union(
+        self,
+        plan: Arc<dyn ExecutionPlan>,
+        children: Vec<RuntimeGroupExec>,
+        config: &ConfigOptions,
+    ) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
+        let handle_children = |children: Vec<RuntimeGroupExec>| -> Result<Arc<dyn ExecutionPlan>> {
+            if children.len() == 1 {
+                self.optimize(children.get(0).unwrap().clone().child, config)
+            } else {
+                Ok(Arc::new(UnionExec::new(
+                    children
+                        .into_iter()
+                        .map(|x| self.optimize(x.child, config))
+                        .collect::<Result<_>>()?,
+                )))
+            }
+        };
+
+        let it = children.into_iter();
+        let mut remote = Vec::new();
+        let mut local = Vec::new();
+
+        for child in it {
+            match child.preference {
+                RuntimePreference::Remote => remote.push(child),
+                RuntimePreference::Local => local.push(child),
+                RuntimePreference::Unspecified => todo!("unspecified runtime preference"),
```

Comment:
I think we should return the untransformed plan in this case instead of panicking. Views are unspecified, so I believe a union between a view and something else would trigger this.

I left views as "unspecified" since the grouping for a view depends on its content (which itself will have the appropriate runtime preference).